plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelsAction.java

```diff
@@ -157,6 +171,94 @@ private void undeployModels(String[] targetNodeIds, String[] modelIds, ActionLis
+                            if (modelWorkNodesBeforeRemoval.get(modelId).length == removedNodeCount) { // undeploy all nodes.
+                                updateDocument.put(MLModel.PLANNING_WORKER_NODES_FIELD, ImmutableList.of());
+                                updateDocument.put(MLModel.PLANNING_WORKER_NODE_COUNT_FIELD, 0);
+                                updateDocument.put(MLModel.CURRENT_WORKER_NODE_COUNT_FIELD, 0);
+                                updateDocument.put(MLModel.MODEL_STATE_FIELD, MLModelState.UNDEPLOYED);
+                            } else { // undeploy partial nodes.
+                                // TODO (to fix) when undeploy partial nodes, the original model status could be partially_deployed,
+                                // and the user could be undeploying not running model nodes, and we should update model status to deployed.
+                                updateDocument.put(MLModel.DEPLOY_TO_ALL_NODES_FIELD, false);
+                                List<String> newPlanningWorkerNodes = Arrays
+                                    .stream(modelWorkNodesBeforeRemoval.get(modelId))
+                                    .filter(x -> !removedNodes.contains(x))
+                                    .collect(Collectors.toList());
+                                updateDocument.put(MLModel.PLANNING_WORKER_NODES_FIELD, newPlanningWorkerNodes);
+                                updateDocument.put(MLModel.PLANNING_WORKER_NODE_COUNT_FIELD, newPlanningWorkerNodes.size());
+                                updateDocument.put(MLModel.CURRENT_WORKER_NODE_COUNT_FIELD, newPlanningWorkerNodes.size());
+                                deployToAllNodes.put(modelId, false);
+                            }
+                            updateRequest.index(ML_MODEL_INDEX).id(modelId).doc(updateDocument);
+                            bulkRequest.add(updateRequest).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+                        }
+                        syncUpInput.setDeployToAllNodes(deployToAllNodes);
+                        ActionListener<BulkResponse> actionListener = ActionListener.runBefore(ActionListener.wrap(bulkResponse -> {
+                            log
+                                .debug(
+                                    "updated model state as undeployed for : {}",
+                                    Arrays.toString(actualRemovedNodesMap.keySet().toArray(new String[0]))
+                                );
+                        }, e -> { log.error("Failed to update model state as undeployed", e); }), context::restore);
+                        client.bulk(bulkRequest, ActionListener.runAfter(actionListener, () -> { syncUpUndeployedModels(syncUpRequest); }));
+                    } else {
+                        syncUpUndeployedModels(syncUpRequest);
+                    }
+                }
+            }
             listener.onResponse(new MLUndeployModelsResponse(r));
```

Comment:
It doesn't seem this can fix the issue, the `listener.onResponse` will be ran ignorance if the `syncUpUndeployedModels` complete or not. Looks like you should change the `syncUpUndeployedModels` method to add a listener and pass the listener from `execute` method to it, and when that successfully returned, invoke the `listener.onResponse`.